nixos/modules/services/networking/ntp/ntpd-rs.nix

```diff
@@ -0,0 +1,60 @@
+{ lib, config, pkgs, ... }:
+
+let
+  cfg = config.services.ntpd-rs;
+  configFile = pkgs.writers.writeTOML "ntpd-rs.toml" cfg.config;
+in
+{
+  options.services.ntpd-rs = {
+    enable = lib.mkEnableOption "Network Time Service (ntpd-rs)";
+    metrics.enable = lib.mkEnableOption "ntpd-rs Prometheus Metrics Exporter";
+    useNetworkingTimeServers = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = lib.mdDoc ''
+        Use source time servers from NixOS option networking.timeServers in config.
+      '';
+    };
+    config = lib.mkOption {
+      type = lib.types.attrs;
+      description = lib.mdDoc "Config rendered to TOML.";
+    };
+  };
+
+  config = lib.mkMerge [
+    (lib.mkIf cfg.enable {
+      environment.systemPackages = [ pkgs.ntpd-rs ];
+      systemd.packages = [ pkgs.ntpd-rs ];
+
+      services.timesyncd.enable = lib.mkForce false;
+      systemd.services.systemd-timedated.environment = { SYSTEMD_TIMEDATED_NTP_SERVICES = "ntpd-rs.service"; };
+
+
+      systemd.services.ntpd-rs = {
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          User = "";
+          Group = "";
+          DynamicUser = true;
+          ExecStart = [ "" "${pkgs.ntpd-rs}/bin/ntp-daemon --config=${configFile}" ];
+        };
+      };
+
+      systemd.services.ntp-rs-metrics = lib.mkIf cfg.metrics.enable {
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          User = "";
+          Group = "";
+          DynamicUser = true;
+          ExecStart = [ "" "${pkgs.ntpd-rs}/bin/ntp-metrics-exporter --config=${configFile}" ];
+        };
+      };
+    })
+    (lib.mkIf (cfg.enable && cfg.useNetworkingTimeServers) {
+      services.ntpd-rs.config.source = lib.flip builtins.map config.networking.timeServers (ts: {
+        mode = "server";
+        address = ts;
+      });
+    })
+  ];
```

Comment:
Merging options hidden behind `cfg.enable` and `cfg.enable && cfg.useNetworkingTimeServers` is just the same as putting the whole block behind `cfg.enable` and the second part behind `cfg.useNetworkingTimeServers`

I've also done a few other things, since it was easier to just edit them in this diff anyway:

- Replaced `services.timesyncd.enable = lib.mkForce false;` with an assertions, explaining why it's bad. Silently overriding options like this is not a good idea IMO.
- Made the `systemd-timedated` environment variable `mkDefault`, as there is no reason for us to force this.
- Removed an empty string in `ExecStart` and just made it a single string.
- Same with the metrics exporter
- The flip is a bit redundant. I think I see what you're going for with it being easier to read the thing you're applying the map to first, but this function is small enough that it just causes more confusion in this case.

```suggestion
  config = lib.mkIf cfg.enable {
    assertions = [
		  {
			  assertion = !config.services.timesyncd.enable;
				message = ''
				  `ntpd-rs` is not compatible with `services.timesyncd`. Please disable one of them.
				'';
			}
		];

    environment.systemPackages = [ pkgs.ntpd-rs ];
    systemd.packages = [ pkgs.ntpd-rs ];

    systemd.services.systemd-timedated.environment = {
		  SYSTEMD_TIMEDATED_NTP_SERVICES = lib.mkDefault "ntpd-rs.service";
		};

    systemd.services.ntpd-rs = {
      wantedBy = [ "multi-user.target" ];
      serviceConfig = {
        User = "";
        Group = "";
        DynamicUser = true;
        ExecStart = "${pkgs.ntpd-rs}/bin/ntp-daemon --config=${configFile}";
      };
    };

    systemd.services.ntp-rs-metrics = lib.mkIf cfg.metrics.enable {
      wantedBy = [ "multi-user.target" ];
      serviceConfig = {
        User = "";
        Group = "";
        DynamicUser = true;
        ExecStart = "${pkgs.ntpd-rs}/bin/ntp-metrics-exporter --config=${configFile}";
      };
    };
    
    services.ntpd-rs.config.source = lib.mkIf cfg.useNetworkingTimeServers (map (ts: {
      mode = "server";
      address = ts;
    }) config.networking.timeServers);
	};
```

Reply:
> Replaced services.timesyncd.enable = lib.mkForce false; with an assertions, explaining why it's bad. Silently overriding options like this is not a good idea IMO.
> Made the systemd-timedated environment variable mkDefault, as there is no reason for us to force this.

This is the same as the `services.chrony` module does it so I just copied the behavior. I will take another look what makes more sense.

> Removed an empty string in ExecStart and just made it a single string.

This is done to override `ExecStart` from the upstream provided service unit.

Reply:
> Merging options hidden behind cfg.enable and cfg.enable && cfg.useNetworkingTimeServers is just the same as putting the whole block behind cfg.enable and the second part behind cfg.useNetworkingTimeServers

This wasn't true with `types.attrs` because no deep merging and evaluation of `lib.mkIf` takes place through the module system.

Reply:
Regarding the assertion: We should actually ensure that only one NTP server is running. We have service module for multiple NTP services. This is not specific to timesyncd but timesyncd is automatically enabled so that's why we need to make sure it is being disabled. The assertion is good, I'll also add it to chrony.